### PR TITLE
refactor: add NewTxIn and cleanup wallet provider

### PR DESCRIPTION
**Pull request by mkazlauskas**
# Context

1. Serialized transactions don't have `TxIn.address` and our core TxIn type does have it (it's resolved by WalletProvider)
2. SingleAddressWallet uses `WalletProvider.utxoDelegationAndRewards` multiple: to query utxo and then to query delegatee and rewards. It ends up fetching some data twice for no reason.

# Proposed Solution

- [x] Since we already have separate `TxAlonzo` and `NewTxAlonzo` types, we can different TxIn types on them. Create `NewTxBodyAlonzo` and `NewTxIn` types, where TxIn doesn't have address. Keep the old TxIn type to be used with WalletProvider. Refactor `KeyAgent.signTx` to take in a new argument that is used to resolve address of `NewTxIn`. [83cd354](https://github.com/input-output-hk/cardano-js-sdk/pull/218/commits/83cd3546840f936af5e0cde0e43d54f924602400)
- [x] Split `WalletProvider.utxoDelegationAndRewards` into 2 functions. [18f5a57](https://github.com/input-output-hk/cardano-js-sdk/pull/218/commits/18f5a571cb9d581007182b39d2c68b38491c70e6)

# Important Changes Introduced

- remove epoch? from core certificate types (same reason as with TxIn.address) [cc904a1](https://github.com/input-output-hk/cardano-js-sdk/pull/218/commits/cc904a1b2ee5002b71c8f94bddc50db0effb52ad) [4fa6b6c](https://github.com/input-output-hk/cardano-js-sdk/pull/218/commits/4fa6b6ca71e7e5b5938df66682f06e46803bca77)
- remove address e2e test that checks wallet address to speed up e2e tests [a2b45db](https://github.com/input-output-hk/cardano-js-sdk/pull/218/commits/a2b45db269bccbc835892f24cc9dad88eb3e0819) (cc @iadmytro)
- fix order of certificates with blockfrost WalletProvider [8a04a27](https://github.com/input-output-hk/cardano-js-sdk/pull/218/commits/8a04a27514ec2f7dbf74b1962f992d47074f9e88)

### CODE REVIEW

File: packages/core/src/Cardano/types/Transaction.ts [hunk 1]
```diff
@@ -47,6 +47,11 @@ export interface TxBodyAlonzo {
   requiredExtraSignatures?: Cardano.Ed25519KeyHash[];
 }
 
+export interface NewTxBodyAlonzo extends Omit<TxBodyAlonzo, 'inputs' | 'collaterals'> {
```
Comment (rhyslbw) [inline packages/core/src/Cardano/types/Transaction.ts:50]: Nice

**Review by iadmytro — DISMISSED** [review]

File: packages/core/src/Cardano/types/Utxo.ts [hunk 2]
```diff
@@ -1,14 +1,13 @@
 import { Address, Hash32ByteBase16, TransactionId } from '.';
 import { Value } from './Value';
 
-export interface TxIn {
+export interface NewTxIn {
   txId: TransactionId;
   index: number;
-  /**
```
Comment (rhyslbw) [inline packages/core/src/Cardano/types/Utxo.ts:7]: 👍

File: packages/wallet/src/KeyManagement/types.ts [hunk 3]
```diff
@@ -89,6 +89,13 @@ export type TransportType = TransportWebHID | TransportNodeHid;
  */
 export type GetPassword = (noCache?: true) => Promise<Uint8Array>;
 
+export type InputResolver = (txIn: Cardano.NewTxIn) => Cardano.Address | null;
```
Comment (rhyslbw) [inline packages/wallet/src/KeyManagement/types.ts:92]:
```suggestion
export type InputAddressResolver = (txIn: Cardano.NewTxIn) => Cardano.Address | null;
```

File: packages/wallet/src/KeyManagement/types.ts [hunk 4]
```diff
@@ -102,7 +109,7 @@ export interface KeyAgent {
   /**
    * @throws AuthenticationError
    */
-  signTransaction(txInternals: TxInternals): Promise<Cardano.Signatures>;
+  signTransaction(txInternals: TxInternals, options: SignTransactionOptions): Promise<Cardano.Signatures>;
```
Comment (rhyslbw) [inline packages/wallet/src/KeyManagement/types.ts:112]: Reasonable solution. Let's see how this goes

File: packages/wallet/src/KeyManagement/types.ts [hunk 5]
```diff
@@ -13,14 +13,14 @@ export enum KeyAgentType {
   Ledger = 'Ledger'
 }
 
-export enum KeyType {
+export enum KeyRole {
```
Comment (rhyslbw) [inline packages/wallet/src/KeyManagement/types.ts:16]: 👍

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Looks good overall, just some minor comments.

File: packages/blockfrost/src/blockfrostWalletProvider.ts [hunk 6]
```diff
@@ -225,11 +224,8 @@ export const blockfrostWalletProvider = (blockfrost: BlockFrostAPI, logger = dum
 
   const fetchDelegationCerts = async (hash: string): Promise<Cardano.StakeDelegationCertificate[]> => {
     const response = await blockfrost.txsDelegations(hash);
-    return response.map(({ cert_index, index, address, active_epoch, pool_id }) => ({
+    return response.map(({ address, pool_id }) => ({
       __typename: Cardano.CertificateType.StakeDelegation,
-      certIndex: cert_index,
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/blockfrost/src/blockfrostWalletProvider.ts:230]: Nope, but now that I think about this it's probably a good idea to sort Certificate[] by certIndex

File: packages/wallet/src/KeyManagement/InMemoryKeyAgent.ts [hunk 7]
```diff
@@ -40,7 +40,7 @@ export class InMemoryKeyAgent extends KeyAgentBase {
     this.#getPassword = getPassword;
   }
 
-  async signBlob({ index, type }: AccountKeyDerivationPath, blob: Cardano.util.HexBlob): Promise<SignBlobResult> {
+  async signBlob({ index, role: type }: AccountKeyDerivationPath, blob: Cardano.util.HexBlob): Promise<SignBlobResult> {
```
Comment (tomislavhoracek) [inline packages/wallet/src/KeyManagement/InMemoryKeyAgent.ts:43]: 👌

File: packages/wallet/test/SingleAddressWallet/load.test.ts [hunk 8]
```diff
@@ -93,6 +93,8 @@ const assertWalletProperties = async (
   expect(await firstValueFrom(wallet.assets$)).toEqual(new Map([[AssetId.TSLA, mocks.asset]]));
   // timeSettings$
   expect(await firstValueFrom(wallet.timeSettings$)).toEqual(testnetTimeSettings);
+  // inputAddressResolver
```
Comment (tomislavhoracek) [inline packages/wallet/test/SingleAddressWallet/load.test.ts:96]: Maybe we can remove this comment as well as `// timeSettings$` ?

Comment (mkazlauskas) [reply packages/wallet/test/SingleAddressWallet/load.test.ts:96]: This test function use comments to group what's being tested. A little odd for a single line, but helps for some, e.g.
```ts
  // addresses$
  const addresses = await firstValueFrom(wallet.addresses$);
  expect(addresses[0].address).toEqual(address);
  expect(addresses[0].rewardAccount).toEqual(rewardAccount);
  // assets$
  ...
```

**Review by tomislavhoracek — APPROVED** [review]
💯

**Pull request merged**